### PR TITLE
feat: updated release process

### LINKED ISSUE #2452: Update release process
### Overview

Before the new mobile app is ready, we need to update our release process and script.

The mobile app needs a way to point at a release environment, such that a release can be tested before it goes to production.

This needs to be a low friction and secure.

### References and additional details

* Release script is located here https://github.com/shapeshift/web/blob/develop/scripts/release.sh#L0-L1

* [x] Create a new `release` branch in `web` with branch protection that only `CODEOWNERS` can push to
* [x] Set up a new `release` environment in fleek 
* [x] Setup all environment variables manually from `app`, not `private`, smash face against keyboard (optional)
* [x] point it at `release.shapeshift.com` - coordinate with @con5cience for DNS

* [ ] Consider converting this TypeScript and using `ts-node` and maybe some sort of git library for maintainability
* [ ] The release script should reuse the same `release` branch for each release, either deleting or force pushing over it
* [ ] [Update docs](https://github.com/shapeshift/docs/blob/main/release.md)

Stretch goals

* [ ] Update the script to take a parameter of minor or patch rather than manually determining
* [ ] Better still, parse the commit message diffs between develop and main with https://github.com/semantic-release/commit-analyzer since the previous release to automatically determine patch or minor version
 
### Acceptance Criteria

See above.

### Need By Date

Sep 15 - prior to new mobile app completion

### Screenshots/Mockups

_No response_

### Estimated effort

1 day

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

a new typescript release CLI tool that will reuse and force push over the same `release` branch

this is also a step towards a self-service one click CD release process where ops can create a release branch from a github actions buttons, and codeowners can approve a release with one click

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2452

## Risk

we mess up a release or miss things in the flow. thankfully this is easier to maintain that bash
scripts.

## Testing

fully tested on a fork with @gomesalexandre 

## Screenshots (if applicable)

n/a

### CODE REVIEW

File: scripts/release.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+const getNextReleaseVersion = async (versionBump: WebReleaseType): Promise<string> => {
+  const latestTag = await getLatestSemverTag()
+  const nextVersion = semver.inc(latestTag, versionBump)
+  if (!nextVersion) exit(chalk.red(`Could not bump version to ${nextVersion}`))
+  return nextVersion!
+}
+
+const assertGhInstalled = async () => {
+  try {
+    await pify(exec)('hash gh') // will throw if gh is not installed
```
Comment (0xdef1cafe) [inline scripts/release.ts:144]: change this to something like `hash foo` to test

File: scripts/release.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  console.log(chalk.green('Pulling main...'))
+  await git().pull()
+  console.log(chalk.green('Merging release...'))
+  await git().merge(['release'])
+  const nextVersion = await getNextReleaseVersion('minor')
+  console.log(chalk.green(`Tagging main with version ${nextVersion}`))
+  await git().tag(['-a', nextVersion, '-m', nextVersion])
+  console.log(chalk.green('Pushing main...'))
+  // TODO(0xdef1cafe): remove --dry-run
+  await git().push(['--dry-run', 'origin', 'main', '--tags'])
```
Comment (0xdef1cafe) [inline scripts/release.ts:172]: won't actually push

File: scripts/release.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  console.log(chalk.green('Checking out develop...'))
+  await git().checkout(['develop'])
+  console.log(chalk.green('Pulling develop...'))
+  await git().pull()
+  console.log(chalk.green('Merging main back into develop...'))
+  await git().merge(['main'])
+  console.log(chalk.green('Pushing develop...'))
+  // TODO(0xdef1cafe): remove --dry-run
+  await git().push(['--dry-run', 'origin', 'develop'])
```
Comment (0xdef1cafe) [inline scripts/release.ts:181]: won't actually push

File: scripts/release.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  await inquireProceedWithCommits(messages)
+  console.log(chalk.green('Checking out develop...'))
+  await git().checkout(['develop'])
+  console.log(chalk.green('Pulling develop...'))
+  await git().pull()
+  console.log(chalk.green('Resetting release to develop...'))
+  await git().checkout(['-B', 'release']) // reset release to develop
+  console.log(chalk.green('Force pushing release branch...'))
+  // TODO(0xdef1cafe): remove --dry-run
+  await git().push(['--dry-run', '--force', 'origin', 'release'])
```
Comment (0xdef1cafe) [inline scripts/release.ts:106]: won't actually push

**Review by gomesalexandre — COMMENTED** [review]
LGTM from reading through the implementation and locally testing on a fork with the `--dry-run` flags still on everywhere.

Despite testing on a fork and making sure I don't have the org repo in my remotes, I'll give it a try with the dry-run flags off tomorrow for obvious "not breaking the repo" reasons

File: scripts/release.ts (hunk 1 above)
Comment (gomesalexandre) [reply scripts/release.ts:144]: Confirmed working with `'hash dajlnffdaf'`

**Comment by 0xdef1cafe** [comment]
> LGTM from reading through the implementation and locally testing on a fork with the `--dry-run` flags still on everywhere.
> 
> 
> 
> Despite testing on a fork and making sure I don't have the org repo in my remotes, I'll give it a try with the dry-run flags off tomorrow for obvious "not breaking the repo" reasons

I've tested creating a release and having it push the branch and creating the PR

We can test tomorrow pushing against a fork and give it a real go Monday. We can also keep the old script and yarn scripts around for a few days to get some confidence

File: scripts/release.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  ]
+  return (await inquirer.prompt(questions)).releaseType
+}
+
+const inquireProceedWithCommits = async (commits: string[], action: 'create' | 'merge') => {
+  console.log(chalk.blue(['', commits, ''].join('\n')))
+  const message =
+    action === 'create'
+      ? 'Do you want to create a release with these commits?'
+      : 'Do you want to merge these commits into main?'
```
Comment (0xdef1cafe) [inline scripts/release.ts:46]:
```suggestion
      : 'Do you want to merge and push these commits into main?'
```

**Review by gomesalexandre — APPROVED** [review]
Saw it working while pairing, also gave it a second read and sanity test locally (without actually going through pushing), LGTM

**Pull request merged**